### ISSUE TITLE
Recommended way to check for Properties 

### ISSUE THREAD

**Issue by ivan-kleshnin** [staltz/cycle#92]
Lets say I have this `Model` for `Picture` component:

```javascript
let Model = Cycle.createModel(Props => ({
  src$: Props.get("src$").startWith("#"),
  title$: Props.get("title$").startWith(""),
  favorite$: Props.get("favorite$").startWith(false),
}));
```

In React we have `propTypes` to declare required properties and their expected types. 
What do you think is the best direction to implement that in Cycle?

**Comment by staltz** [comment]
Props is a DataFlowNode, and any actual values are hidden under some layers: DataFlowNode has Observables which have values. So any validation of property types has to be done for instance like this `Props.get('title$').map(checkStringOrThrow)`
 
That said, JavaScript is a dynamically typed language, and what React attempts witb propTypes is to bring in more type safety as a convenience. It's not core to what React does, and neither is to Cycle, so I don't think it should be core to the framework. There are plenty of tools to assist in this aspect. Like TypeScript, FlowType, and other smaller projects. Actually I am considering myself to use TypeScript + Cycle + Immutable.js + JSX in an upcoming project at work.
 
TL;DR: TypeScript

**Comment by ivan-kleshnin** [comment]
Thanks! I considered `tap` and `map` as obvious tools but was not 100% sure.

**Comment by staltz** [comment]
PS: tap is deprecated, the correct one is doOnNext()

**Comment by ivan-kleshnin** [comment]
I don't see deprecation note [here](https://github.com/Reactive-Extensions/RxJS/blob/master/doc/api/core/operators/do.md) :confused: 

Anyway I greatly appreaciate and hopefully share philosophy behind your answers:
> Choose right tool for the task (fake quote)

React PropTypes are kind of hack, because they force you to validate state with one tool, props with second and types with third. Three tools == three different syntaxes to remember. It's again about "magic" vs "boilerplate".

**Comment by staltz** [comment]
Ah, so tap is not deprecated, sorry about that

**Comment by cgeorg** [comment]
If you wanted some sort of type checking, it would be easy enough to create
a new DataFlowNode that performed type checking via tap.  You could then
inject it between any 2 other nodes depending on environment.

The same could really be done for any sort of AOP - add logging, etc.

**Comment by staltz** [comment]
Closing this issue, will reopen in case you want to continue the discussion.

**Comment by tel** [comment]
I'm excited by how amenable Cycle is to Typescript. For people who are interested I have [a declaration file for virtual-dom](https://github.com/tel/typescript-vdom/blob/master/reference/virtual-dom.d.ts) and am working on [a partial reimplementation of virtual-dom in Typescript](https://github.com/tel/typescript-vdom).

**Comment by ivan-kleshnin** [comment]
@tel I'm very interested in TypeScript, but have little time to play around. How exactly do you use it: is it compatible with Babel and ES6 syntax? Webpack?

**Comment by tel** [comment]
@ivan-kleshnin TypeScript is a bit tough to get started on and it's certainly not 100% compatible with ES6. Frankly, in my own use, what TypeScript brings is far more valuable than what ES6 brings, but that's pure opinion.

You can use JSX with Typescript, but you'll lose editor integration (or rather just suffer a lot of complaints from your editor) as you'll need a pipeline which desugars your JSX prior to hitting `tsc`.

Practically, I'm avoiding JSX (I've never much liked it, tbh) and using [a gulp workflow involving `tsc` and `browserify`](https://gist.github.com/tel/ac24127a13540122813d). It took a little bit of time to get set up, but seems to work alright. The only outstanding difficulty is that the `watchify` runner crashes on TypeScript errors. In practice this means that I avoid saving files until I've eliminated any obvious type errors... probably will become a larger problem in time, but doesn't slow things down appreciably as of now.

I've tried Flow a few times. I found it less mature, but I'd be tempted to use it exactly and only for its treatment of `null`s. They are such a terrible thing to have in a language.

**Comment by ivan-kleshnin** [comment]
@tel thank you very much for such detailed explanation. 

> I've tried Flow a few times. I found it less mature, but I'd be tempted to use it exactly and only for its treatment of nulls. They are such a terrible thing to have in a language. 

As I remember from theory it's not a `null` on itself being terrible but default "nullability" of all types in JS / Java. Anyway, it's an interesting point in Flow.

> I think ultimately Flow wants to be a type-checking layer atop JavaScript while TypeScript wants to be a new language which just so happens to superset JavaScript. I feel the latter is a better place to stand.

New language makes a difference, yes. There are pros and cons. I would rather go with ClojureScript, I'm really tired of JS, but there are a lot more libraries, a lot more working opportunities and some super-valueable tools like Webpack are unique to JS ecosystem.

**Comment by tel** [comment]
> default "nullability" of all types in JS / Java

Yeah, that's one way to see it. The technical form I like the most is the notion that all types are "pointed" in that they all contain a special value "null". It turns out that it's rare to have semantics amenable to this treatment (but consider mathematical groups which have "null" but handle it gracefully).
